R/fit_trait_distributions.R

```diff
@@ -0,0 +1,130 @@
+
+#' @examples
+#' data(community)
+#' data(trait)
+#' imputed_traits <-trait_impute(comm = community, traits = trait,
+#'                  scale_hierarchy = c("Site", "PlotID"),
+#'                  taxon_col = "Taxon", value_col = "Value",
+#'                  trait_col = "Trait", abundance_col = "Cover")
+#' fitted_distributions <- fit_trait_distributions(
+#' imputed_traits = imputed_traits,
+#' distribution_type = "normal")
+#'
+#' @export
+fit_trait_distributions <- function(imputed_traits,
+                                    distribution_type = "normal") {
+
+  #Pull useful information from imputed traits object
+  value_col <- attributes(imputed_traits)$attrib$value_col
+  trait_col <- attributes(imputed_traits)$attrib$trait_col
+  taxon_col <- attributes(imputed_traits)$attrib$taxon_col
+  abundance_col <- attributes(imputed_traits)$attrib$abundance_col
+  scale_hierarchy <- attributes(imputed_traits)$attrib$scale_hierarchy
+
+  #If only a single type of distribution was supplied, apply it to all traits
+  if (length(distribution_type) == 1) {
+    distributions <- unique(as.data.frame(imputed_traits)[, trait_col])
+    dist_list <- replicate(n = length(distributions), expr = distribution_type)
+    names(dist_list) <- distributions
+    distribution_type <- as.list(dist_list)
+    rm(dist_list, distributions)
+  } #if statement
+
+  #Data checks
+
+  #check distribution type is supported
+  if (any(!unique(distribution_type) %in% c("beta", "normal", "lognormal"))) {
+    x <- distribution_type[!distribution_type %in%
+                            c("beta", "normal", "lognormal")]
+    stop(glue("distribution type '{x}' not supported "))
+  }
+
+
+  #Beta checks
+  if ("beta" %in% distribution_type) {
+
+    #make sure every species x hierarchy combination has at least 2 data points
+    beta_counts <- imputed_traits %>%
+      filter(.data[[trait_col]] ==
+               names(distribution_type)[distribution_type == "beta"]) %>%
+      group_by_at(c(as.character(scale_hierarchy), trait_col, taxon_col)) %>%
+      summarise(n = n(), .groups = "keep")
+
+    if (any(beta_counts$n < 2)) {
+    stop("Fitting a beta distrbution requires 2+ points per distribution.
+         We suggest re-imputing traits with a minimum of (at least) 2 leaves.")
+      }
+
+    #check that values are between 0 and 1
+    beta_vals <- imputed_traits %>%
+      ungroup() %>%
+      filter(.data[[trait_col]] ==
+               names(distribution_type)[distribution_type == "beta"]) %>%
+      select(all_of(value_col))
+
+    if (any(beta_vals > 1 | beta_vals < 0)) {
+      stop("For a Beta distribution values must be between 0 and 1.")
+    }
+
+
+  }# end beta checks
+
+
+  #lognormal checks
+  if ("lognormal" %in% distribution_type) {
+
+    ln_vals <- imputed_traits %>%
+      ungroup() %>%
+      filter(.data[[trait_col]] ==
+               names(distribution_type)[distribution_type == "lognormal"]) %>%
+      dplyr::select(all_of(value_col))
```

Comment:
you don't need dplyr::, just put it in the @importFrom.

Reply:
Fixed.